consensus/qkchash/algorithm_native.go

```diff
@@ -0,0 +1,111 @@
+//+build nt
+
+package qkchash
+
+import (
+	"encoding/binary"
+	"sort"
+	"sync"
+
+	"github.com/QuarkChain/goquarkchain/consensus/qkchash/native"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+const (
+	cacheEntryCnt   = 1024 * 64
+	cacheAheadRound = 64 // 64*30000
+)
+
+var (
+	EpochLength = uint64(30000) //blocks pre epoch
+)
+
+type cacheSeed struct {
+	mu     sync.RWMutex
+	caches []qkcCache
+}
+
+func NewcacheSeed() *cacheSeed {
+	firstCache := generateCache(cacheEntryCnt, common.Hash{}.Bytes())
+	caches := make([]qkcCache, 0)
+	caches = append(caches, firstCache)
+	return &cacheSeed{
+		caches: caches,
+	}
+}
+
+func (c *cacheSeed) getCacheFromHeight(block uint64) qkcCache {
+	epoch := int(block / EpochLength)
+	lenCaches := len(c.caches)
+	if epoch < lenCaches {
+		return c.caches[epoch]
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	needAddCnt := epoch - lenCaches + cacheAheadRound
+	seed := c.caches[len(c.caches)-1].seed
+	for i := 0; i < needAddCnt; i++ {
+		seed = crypto.Keccak256(seed)
+		c.caches = append(c.caches, generateCache(cacheEntryCnt, seed))
```

Comment:
so we never remove cache? should we trim the old entries (like LRU-style) avoid consuming too much memory?

Reply:
One cache for every 30000 blocks, should not cache too much.

Reply:
that's about 4~5 days. if we want to make the process long-lasting, should we at least trim the cache list a little bit?

Reply:
Would this mean that we will cache for all epochs?  E.g., 3720294 miner blocks means 124 caches?  BTW: The code here seems to be duplicate of non-native version.

Reply:
3720294 miner blocks mean 124 caches? yes. so 16k for one year.
The code here seems to be duplicate of non-native version. I have removed dup code.

Reply:
after test, in one cache, ls use 1m, and nativeCache use 2m. add remove cache code.